cli/tests/test_diff_command.rs

```diff
@@ -268,6 +268,146 @@ fn test_diff_empty() {
+    std::fs::write(repo_path.join("file2"), "1\n").unwrap();
+    std::fs::write(repo_path.join("file3"), "1\n").unwrap();
+    std::fs::write(repo_path.join("file4"), "").unwrap();
+    test_env.jj_cmd_ok(&repo_path, &["file", "chmod", "x", "file1", "file2"]);
+
+    test_env.jj_cmd_ok(&repo_path, &["new"]);
+    std::fs::write(repo_path.join("file1"), "2\n").unwrap();
+    std::fs::write(repo_path.join("file3"), "2\n").unwrap();
+    test_env.jj_cmd_ok(&repo_path, &["file", "chmod", "n", "file1", "file2"]);
+    test_env.jj_cmd_ok(&repo_path, &["file", "chmod", "x", "file3", "file4"]);
+
+    test_env.jj_cmd_ok(&repo_path, &["new"]);
+    std::fs::remove_file(repo_path.join("file1")).unwrap();
+    std::fs::remove_file(repo_path.join("file2")).unwrap();
+    std::fs::remove_file(repo_path.join("file3")).unwrap();
+    std::fs::remove_file(repo_path.join("file4")).unwrap();
+
+    let stdout = test_env.jj_cmd_success(&repo_path, &["diff", "-r@--"]);
+    insta::assert_snapshot!(stdout, @r###"
+    Added executable file file1:
+        (empty)
+    Added executable file file2:
+            1: 1
+    Added regular file file3:
+            1: 1
+    Added regular file file4:
+        (empty)
+    "###);
+    let stdout = test_env.jj_cmd_success(&repo_path, &["diff", "-r@-"]);
+    insta::assert_snapshot!(stdout, @r###"
+    Executable file became non-executable at file1:
+            1: 2
+    Executable file became non-executable at file2:
+    Non-executable file became executable at file3:
+       1    1: 12
+    Non-executable file became executable at file4:
+    "###);
+    let stdout = test_env.jj_cmd_success(&repo_path, &["diff", "-r@"]);
+    insta::assert_snapshot!(stdout, @r###"
+    Removed regular file file1:
+       1     : 2
+    Removed regular file file2:
+       1     : 1
+    Removed executable file file3:
+       1     : 2
+    Removed executable file file4:
+        (empty)
+    "###);
+
+    let stdout = test_env.jj_cmd_success(&repo_path, &["diff", "-r@--", "--git"]);
+    insta::assert_snapshot!(stdout, @r###"
+    diff --git a/file1 b/file1
+    new file mode 100755
+    index 0000000000..e69de29bb2
+    diff --git a/file2 b/file2
+    new file mode 100755
+    index 0000000000..d00491fd7e
+    --- /dev/null
+    +++ b/file2
+    @@ -1,0 +1,1 @@
+    +1
+    diff --git a/file3 b/file3
+    new file mode 100644
+    index 0000000000..d00491fd7e
+    --- /dev/null
+    +++ b/file3
+    @@ -1,0 +1,1 @@
+    +1
+    diff --git a/file4 b/file4
+    new file mode 100644
+    index 0000000000..e69de29bb2
+    "###);
+    let stdout = test_env.jj_cmd_success(&repo_path, &["diff", "-r@-", "--git"]);
+    insta::assert_snapshot!(stdout, @r###"
+    diff --git a/file1 b/file1
+    old mode 100755
+    new mode 100644
+    index e69de29bb2..0cfbf08886
+    --- a/file1
+    +++ b/file1
+    @@ -1,0 +1,1 @@
+    +2
+    diff --git a/file2 b/file2
+    old mode 100755
+    new mode 100644
+    diff --git a/file3 b/file3
```

Comment:
Do I understand correctly that this is a bug and it is not fixed in this PR?

If so, can you add a comment explaining what the bug is?

Reply:
No, it's not a bug (if the quoted region is displayed correctly.) You can see header-only unified diffs in 5e2a9c274698f4b69b4b58cf226a811edd166e47.

The goal is to disable diff of binary files. I just found a couple of bugs while moving codes around, and I needed some tests covering these.